### ISSUE TITLE
Allow dictionary as second param in WebSocket constructor

### ISSUE THREAD

**Issue by KhafraDev** [whatwg/websockets#42]
This would allow server environments (specifically talking about nodejs/undici) that implement WebSocket some flexibility, for example, to add custom headers, etc. While we could change it on our side, it would be incompatible with other implementations. Similarly, it'll cause type issues with people who use typescript in cross-env packages (ie. https://github.com/nodejs/undici/issues/1943).

Proposed change:
```webidl
dictionary WebSocketInit {
  protocols: DOMString or sequence<DOMString>
}

interface WebSocket : EventTarget {
  constructor(USVString url, optional (DOMString or sequence<DOMString> or WebSocketInit) options = []);
```

**Comment by ricea** [comment]
It's a sensible extension point and as far as I know compatible with existing practice. But it's hard to motivate it from a browser implementation point of view.

In other words, I'm not opposed, but I don't know how I'd justify the work.

**Comment by KhafraDev** [comment]
understandable, I was hoping it would be minor enough for browser implementers to not mind lol

**Comment by KhafraDev** [comment]
Bun also allows a dictionary as second parameter for setting custom headers, example [here](https://bun.sh/docs/api/websockets#connect-to-a-websocket-server).

**Comment by KhafraDev** [comment]
it was also suggested in https://github.com/whatwg/websockets/issues/16#issuecomment-881547333 which I didn't know about

**Comment by pimterry** [comment]
Worth noting that the _extremely_ popular `ws` package (now 110 million downloads a week - more than `react` + `express` combined!) supports 3 arguments ([docs](https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketaddress-protocols-options)) as does `faye-websocket` (20 million/week, [docs](https://www.npmjs.com/package/faye-websocket#initialization-options)). In this case the protocol is optional instead of just overloading the 2nd arg:

```js
new WebSocket(address[, protocols][, options])
```

In that case, `protocol` isn't included as a valid option in `options` - if you want to specify subprotocols, you have to provide all 3 arguments. That protocols argument is optional though, so if you don't need it, you can use `new WebSocket(address, options)` with those libraries as well.

This API has the advantage of supporting out-of-the-box browser compatibility: `new WebSocket(url, { ... })` will throw in browsers due to the invalid argument, so any code using that API is not currently portable to browsers (and probably never will be), while `new WebSocket(url, null, { ... })` works fine in browsers today, with the options simply being ignored.

Standardizing options as a 3rd argument would help to ensure browsers preserve that status quo, which means we'd have a single API offering easy compatibility everywhere, with minimal effort for everybody. Lovely.

Would be nice to see some consensus on the different approaches here, since a consistent API would be very helpful. @KhafraDev @mcollina @uzlopak on the Undici side, would you be open to matching the `ws` API here, and starting to gently move away from your current websocket options API? Looks very easy to support both without any breaking changes. I'm not sure who to ping for Bun but hopefully they'd be open to following suit. CC @lpinca who might also have opinions here.

**Comment by annevk** [comment]
I think if we are to do this (and it looks like we might have to for Local Network Access) the shape proposed in OP makes sense. We generally don't go for an optional argument in the middle.

**Comment by pimterry** [comment]
@annevk Fair enough - if browsers are indeed interested in broader options support as well, that would make plenty of sense (which would imply `ws` et al changing to support that instead, but that migration is also easy and non-breaking so not a huge problem).

It'd need some form of feature detection for browsers though, since `new WebSocket(url, { targetAddressSpace: 'private' })` throws in at least Chrome & FF today, so it wouldn't be safe for people to start using that API even if it existed. I'm not sure what the consensus is on approach for those kind of changes in browsers nowadays, but from my POV anyway (on the Node side) the end goal sounds fine.

**Comment by annevk** [comment]
Throwing is generally perfect for feature detection, no?

**Comment by pimterry** [comment]
:shrug: that'll certainly work. Mildly messy but perfectly practical. I don't know what the normal web dev expectations for incompatible API changes like this are.

**Comment by crowlKats** [comment]
for information, Deno has added in a recent release to accept an object as a secondary parameter, which looks the following:
```ts
interface WebSocketOptions {
  protocols?: string | string[];
  headers?: HeadersInit;
}
```
behind unstable as it was getting quite a problem for us (both from users perspective and for our needs internally), hoping that if the spec adds this it would look something similar to this.

**Comment by KhafraDev** [comment]
> for information, Deno has added in a recent release to accept an object as a secondary parameter, which looks the following:

This is identical to what node supports. We haven't had any reports of issues regarding it, but it's rather obscure.

**Comment by ricea** [comment]
Blink will also need this for `targetAddressSpace`. That means it is more likely to happen now.

**Comment by christhompson** [comment]
> Blink will also need this for `targetAddressSpace`. That means it is more likely to happen now.

Yes, we are now starting to look into this with a higher priority as one of the requirements for being able to be strict about requiring secure contexts for Local Network Access. Do let us know if we can help drive this forward.

**Comment by christhompson** [comment]
I've opened PR #76 to update the spec to support this (based on the original API design of a single optional second param that can now _also_ be a WebSocketInit dictionary). I also have a [draft implementation for this in Chromium](https://crrev.com/c/crrev.com/c/8197225) that adds WPTs.
